### ISSUE TITLE
Source the direct-RL zero-advantage collapse claim with a public ablation run

### ISSUE THREAD

**Issue by tokenbender** [mananchawla2005/adaptive-tokenization#85]
## Summary

The README claims that direct RL on the boundary predictor was tried and collapsed into zero advantage, while the cold-started GRPO run kept advantage high in the attached W&B logs.

The public W&B project currently exposes eight runs, and none is a direct-RL ablation. The only GRPO run is the cold-started Stage 3 run from the hybrid Phase 2 pipeline. The checked-in direct-RL implementation in `src/grpo_train.py` also has no Modal entrypoint in the README path and no W&B tracker.

So the negative baseline claim may be true, but it is not backed by the visible public evidence surface.

This is distinct from #56 and #41:

- #56 covers whether the logged cold-started `advantages_mean` metric measures raw reward spread.
- #41 covers the stale/incompatible direct-GRPO implementation left in `src/`.
- This issue covers the missing public ablation evidence for the README claim that direct RL collapsed.

## Evidence in this repo

Current commit checked: `d53f400df338eee48641df72ea9aad2a6950fd14`.

1. README claims direct RL collapsed, and says the contrast is visible in attached W&B logs:
   - https://github.com/mananchawla2005/adaptive-tokenization/blob/d53f400df338eee48641df72ea9aad2a6950fd14/README.md#L63-L65

2. README reproduction commands expose only the hybrid Phase 2 path, not a direct-RL ablation command:
   - https://github.com/mananchawla2005/adaptive-tokenization/blob/d53f400df338eee48641df72ea9aad2a6950fd14/README.md#L120-L132

3. `phase2_app.py` imports and runs `stage3_train_grpo` from `src.hybrid_train`, after Stage 1 oracle creation and Stage 2 BCE imitation:
   - https://github.com/mananchawla2005/adaptive-tokenization/blob/d53f400df338eee48641df72ea9aad2a6950fd14/phase2_app.py#L47-L58
   - https://github.com/mananchawla2005/adaptive-tokenization/blob/d53f400df338eee48641df72ea9aad2a6950fd14/phase2_app.py#L119-L134

4. The standalone direct-RL-like implementation is `src/grpo_train.py::train_grpo`, but it is not imported by the Modal app or README reproduction path:
   - https://github.com/mananchawla2005/adaptive-tokenization/blob/d53f400df338eee48641df72ea9aad2a6950fd14/src/grpo_train.py#L100-L216

5. That standalone implementation also has no `Tracker`/W&B logging, so it cannot be the public W&B evidence for a zero-advantage collapse as written:
   - https://github.com/mananchawla2005/adaptive-tokenization/blob/d53f400df338eee48641df72ea9aad2a6950fd14/src/grpo_train.py#L100-L216

## Public W&B evidence

Unauthenticated W&B GraphQL run listing for `highattacker/adaptive-tokenization` currently returns eight runs:

```text
exp-20260623-144000-gd53f400-phase2-eval
exp-20260623-131726-gd53f400-phase2-stage3-grpo
exp-20260623-131726-gd53f400-phase2-stage2-bce
exp-20260623-131726-gd53f400-phase2-stage1-oracle
exp-20260622-103522-g2e375ab-oracle
exp-20260622-100551-g2e375ab-phase1-eval
exp-20260622-075740-g2e375ab-phase1-adaptive
exp-20260622-075734-g2e375ab-phase1-naive
```

The only GRPO training run is `xhip54h0`, whose display name is `exp-20260623-131726-gd53f400-phase2-stage3-grpo`:

- https://wandb.ai/highattacker/adaptive-tokenization/runs/xhip54h0

That is the post-BCE, cold-started Stage 3 run. There is no visible direct-RL baseline run, no `job_type=direct-grpo`, no pre-BCE GRPO run, and no run whose display name indicates a direct-RL ablation.

## Why this is a bug / expected behavior

A negative ablation claim needs the same provenance standard as a positive result. If the README says direct RL collapsed and points readers to W&B logs, the public project should contain a clearly named run or artifact showing:

- direct-RL setup,
- checkpoint/config,
- reward/advantage metrics,
- collapse criterion,
- seed and dataset/sample protocol,
- comparison to the cold-started pipeline.

Without that, the claim reads as a conclusion without an auditable experiment.

## Suspected culprit

The direct-RL experiment may have been run locally or in an earlier private/ad hoc setting, while the README was updated to cite public W&B logs that only contain the final hybrid pipeline.

### nanochat

nanochat writes eval outputs and report sections as artifacts so comparisons are tied to generated evidence:

- https://github.com/karpathy/nanochat/blob/dc54a1a3077cab11d68fac4c5d1cd5c51f5d8c7a/scripts/base_eval.py#L285-L317

Its RL script also logs reward and loss metrics from the actual RL run:

- https://github.com/karpathy/nanochat/blob/dc54a1a3077cab11d68fac4c5d1cd5c51f5d8c7a/scripts/chat_rl.py#L279-L294

### TorchTitan

TorchTitan treats training/evaluation state as configured run metadata and logs validation metrics from that configured path:

- https://github.com/pytorch/torchtitan/blob/6392a5da3a349cb402ed8bc2a063382abe2e6618/torchtitan/trainer.py#L181-L195
- https://github.com/pytorch/torchtitan/blob/6392a5da3a349cb402ed8bc2a063382abe2e6618/torchtitan/components/metrics.py#L539-L563

These references establish the expected discipline: method-comparison claims should have named run artifacts for each compared method.

## Suggested fix

- Add a direct-RL ablation entrypoint with the same tracker/provenance metadata as the hybrid pipeline.
- Log direct-RL reward spread, raw advantage, normalized advantage, compression, loss, seed, checkpoint, and commit id.
- Link the exact direct-RL W&B run/artifact from the README.
- If the ablation artifact cannot be recovered, soften the README to say direct RL was an informal attempt rather than a logged result.

## Expected impact

The current README may overstate the evidence for the three-stage hybrid design. The positive BCE+GRPO result is public, but the negative direct-RL baseline that motivates cold starting is not publicly auditable.
